src/util/file_writer.js

```diff
@@ -0,0 +1,140 @@
+const stream = require('stream');
+const config = require('../../config');
+const nb_native = require('./nb_native');
+const dbg = require('../util/debug_module')(__filename);
+
+/**
+ * FileWriter is a Writable stream that write data to a filesystem file,
+ * with optional calculation of md5 for etag.
+ */
+class FileWriter extends stream.Writable {
+
+    /**
+     * @param {{
+     *      target_file: object,
+     *      fs_context: object,
+     *      namespace_resource_id: string,
+     *      md5_enabled: boolean,
+     *      stats: import('../sdk/endpoint_stats_collector').EndpointStatsCollector,
+     *      offset?: number,
+     *      bucket?: string,
+     *      large_buf_size?: number,
+     * }} params
+     */
+    constructor({ target_file, fs_context, namespace_resource_id, md5_enabled, stats, offset, bucket, large_buf_size }) {
+        super({ highWaterMark: config.NSFS_UPLOAD_STREAM_MAX_BUFFER_SIZE });
+        this.target_file = target_file;
+        this.fs_context = fs_context;
+        this.offset = offset;
+        this.total_bytes = 0;
+        this.count_once = 1;
+        this.stats = stats;
+        this.bucket = bucket;
+        this.namespace_resource_id = namespace_resource_id;
+        this.large_buf_size = large_buf_size || config.NSFS_BUF_SIZE_L;
+        this.MD5Async = md5_enabled ? new (nb_native().crypto.MD5Async)() : undefined;
+        const platform_iov_max = nb_native().fs.PLATFORM_IOV_MAX;
+        this.iov_max = platform_iov_max ? Math.min(platform_iov_max, config.NSFS_DEFAULT_IOV_MAX) : config.NSFS_DEFAULT_IOV_MAX;
+    }
+
+    /**
+     * @param {number} size 
+     */
+    _update_stats(size) {
+        const count = this.count_once;
+        this.count_once = 0; // counting the entire operation just once
+        this.stats?.update_nsfs_write_stats({
+            namespace_resource_id: this.namespace_resource_id,
+            size,
+            count,
+            bucket_name: this.bucket,
+        });
+    }
+
+    /**
+     * @param {Buffer[]} buffers 
+     * @param {number} size 
+     */
+    async _update_md5(buffers, size) {
+        // TODO optimize by calling once with all buffers
+        for (const buf of buffers) {
+            await this.MD5Async.update(buf);
+        }
+    }
+
+    /**
+     * @param {Buffer[]} buffers 
+     * @param {number} size 
+     */
+    async _write_all_buffers(buffers, size) {
+        if (buffers.length <= this.iov_max) {
+            await this._write_to_file(buffers, size);
+        } else {
+            let iov_start = 0;
+            while (iov_start < buffers.length) {
+                const iov_end = Math.min(buffers.length, iov_start + this.iov_max);
+                const buffers_to_write = buffers.slice(iov_start, iov_end);
+                const size_to_write = buffers_to_write.reduce((s, b) => s + b.length, 0);
+                await this._write_to_file(buffers_to_write, size_to_write);
+                iov_start = iov_end;
+            }
```

Comment:
since this case is very rare, and I don't know if our tests will easily get to it, I would like to suggest that in our unit tests we explicitly set iov_max to a low value such as 1 or 2 and make sure we test also this case.

Reply:
I added a test a few years ago for this iov_max case - https://github.com/noobaa/noobaa-core/pull/6928/files#diff-a5e4de68b0630a9cd0aab52d062a381429b3e890302696803ceadc2b1acb552aR22-R33
but anyway, I added in test_file_writer 4 new cases with your suggestion.